### ISSUE TITLE
[META] Use automated testing / Add .deb/.rpm/pacman packages

### ISSUE THREAD

**Issue by ehfd** [selkies-project/selkies#58]
- Build and add .deb and .rpm packages (help for Arch or Alpine distros also welcome)

We should definitely do this, but we need help. This requires build procedures for every relevant distro.

- Testing workflows

In every PR, we should test using Selenium and Firefox an end-to-end procedure, starting from container builds to testing that GStreamer starts, and we see that the Start button of the web interface is visible.

- Formatting, automatic code assessment tools, and a single syntax for code documentation

We should incorporate format checking such as `pylint` or `yapf` and code testing. Similarly for JavaScript. Code checking and assessments should automatically be done when a Pull Request starts.
People with relevant experience are welcome to reach out.

Code documentation should also be added to the code itself. There are well-documented locations and those that aren't really so.

These should be done before the codebase becomes substantially larger than now.

**Comment by ehfd** [comment]
https://github.com/pacstall/pacstall
I also stumbled upon this; an AUR for Ubuntu?

**Comment by ehfd** [comment]
@danisla 
The following are outdated actions which require major updates to the pipeline:
[`tj-actions/changed-files@v1.1.2`](https://github.com/tj-actions/changed-files) (newest version does not work just by changing the tag)
[`actions/create-release@v1`](https://github.com/actions/create-release) (archived and deprecated)
[`actions/upload-release-asset@v1`](https://github.com/actions/upload-release-asset)  (archived and deprecated)

**Comment by ehfd** [comment]
@danisla The Actions pipeline is buggy. The release for GStreamer is always botched for some reason when `Build & publish all images` is not run before `Publish release`, and the actions are outdated.

Ref: https://github.blog/changelog/2022-10-11-github-actions-deprecating-save-state-and-set-output-commands/

**Comment by ehfd** [comment]
@danisla It seems that the release pipeline is broken. GStreamer is not delivered correctly especially in Ubuntu 20.04 and apparently breaks something with the caching mechanism.

For now, upload releases manually.

**Comment by danisla** [comment]
> @danisla It seems that the release pipeline is broken. GStreamer is not delivered correctly especially in Ubuntu 20.04 and apparently breaks something with the caching mechanism.
> 
> For now, upload releases manually.

The pipeline fixes should be a separate bug, adding .deb packages is a new feature.

**Comment by ehfd** [comment]
@danisla I won't open a new issue because I feel the CI is nearly fixed, but is it intended that gst-py-examples containers are not built in "Build & publish changed images"?

**Comment by danisla** [comment]
I was attempting to only build the gst-py-examples if any of the matrix tasks changed, but it's not obvious that this is possible ATM.
This commit should always build the example images: https://github.com/selkies-project/selkies-gstreamer/commit/587ec412e9a64f5146bc06abb9c77f1f2c1dde3d

**Comment by danisla** [comment]
@ehfd for new releases, the process has changed slightly. The pipeline no longer creates the GH release.

1. manually create the Github Release and Tag.
2. Generate the release notes with the github button.
3. The pipeline will trigger once the release is published and will upload the assets to the release.

**Comment by ehfd** [comment]
Default branch was changed to `main` to comply with GitHub guidances. If you have a local clone, please use:

```
git branch -m master main
git fetch origin
git branch -u origin/main main
git remote set-head origin -a
```

**Comment by ehfd** [comment]
> @ehfd for new releases, the process has changed slightly. The pipeline no longer creates the GH release.
> 
> 1. manually create the Github Release and Tag.
> 2. Generate the release notes with the github button.
> 3. The pipeline will trigger once the release is published and will upload the assets to the release.

I have complied to this process. Best way is to set the new release as a pre-release until the assets are uploaded then change to latest.

**Comment by danisla** [comment]
> Default branch was changed to `main` to comply with GitHub guidances. If you have a local clone, please use:
> 
> ```
> git branch -m master main
> git fetch origin
> git branch -u origin/main main
> git remote set-head origin -a
> ```

Are you going to update all the things that depend on `master`?

**Comment by ehfd** [comment]
> > Default branch was changed to `main` to comply with GitHub guidances. If you have a local clone, please use:
> > 
> > ```
> > git branch -m master main
> > git fetch origin
> > git branch -u origin/main main
> > git remote set-head origin -a
> > ```
> 
> Are you going to update all the things that depend on `master`?

I am doing it.

**Comment by ehfd** [comment]
> I am doing it.

Done.

I am freezing both the desktop containers and Selkies-GStreamer after v1.4.3 and withdrawing for a few months, unless immediate issues arise.

Please ask me before a release, as it can affect the desktop containers.

**Comment by danisla** [comment]
Not sure if I’m ok freezing all Selkies gstreamer releases for a few months. Can you lock the desktop containers to Selkies-gstreamer v1.4.3? Having them roll with latest doesn’t seem like good CI practice anyway.

**Comment by ehfd** [comment]
> Not sure if I’m ok freezing all Selkies gstreamer releases for a few months. Can you lock the desktop containers to Selkies-gstreamer v1.4.3? Having them roll with latest doesn’t seem like good CI practice anyway.

If so, you can make a release, no problem.
But I still do want to roll with latest.
I'll manually trigger a rebuild when I can check things. That way the container images will be alive even after releases introduce incompatibilities.

**Comment by ehfd** [comment]
Unless something is very out of the ordinary, the container desktops will be kept as-is for at least a few months (only Selkies-GStreamer dependencies will be updated as there are new releases).
I have exhausted and fixed every reason the container might break.

@danisla You are free to make releases, just contact and tell me when you have changed the GStreamer Dockerfile (or its runtime dependencies in Dockerfile.example). Anything otherwise will not interfere with the containers.

**Comment by ehfd** [comment]
@danisla
It seems a good idea to:
1. Create a merged .deb package which includes the Python, GStreamer, Web, and the Interposer.
2. Ship a .tgz archive of the interposer .so file, because one of the advantages of using Selkies is not having to install as root.
3. How would it be possible to create a completely portable version of all the components?

**Comment by ehfd** [comment]
@danisla `build_changed_images` is broken, looks like it's incomplete.

https://github.com/selkies-project/selkies-gstreamer/blob/2d697395f0f5fc6c3f0db2ea7c08d26e0594259b/.github/workflows/build_changed_images.yaml#L98-L133

**Comment by ehfd** [comment]
TODO: Change https://github.com/selkies-project/selkies-gstreamer/blob/main/.github/actions/build_and_publish_image/action.yaml with https://github.com/docker/build-push-action.

**Comment by ehfd** [comment]
Remaining tasks:
Build and add .deb and .rpm packages (help for Arch or Alpine distros also welcome): This is still preferred despite considering Conda and AppImages.
Formatting, automatic code assessment tools, and a single syntax for code documentation: Consider SonarCube as a potential possibility.

**Comment by ehfd** [comment]
Fixed in commits including https://github.com/selkies-project/selkies/commit/123fe20bd71ced396295e935d35ef52a2dd2f57e.
